File: python_bot/bot.py
import asyncio
import logging
import os
import sqlite3
from datetime import datetime
from dotenv import load_dotenv

from aiogram import Bot, Dispatcher, F, types
from aiogram.filters import CommandStart, CommandObject
from aiogram.fsm.context import FSMContext
from aiogram.fsm.state import State, StatesGroup
from aiogram.types import InlineKeyboardMarkup, InlineKeyboardButton

load_dotenv()
BOT_TOKEN = os.getenv("BOT_TOKEN")
ADMIN_IDS = [int(id) for id in os.getenv("ADMIN_IDS", "").split(",") if id]

logging.basicConfig(level=logging.INFO)

bot = Bot(token=BOT_TOKEN)
dp = Dispatcher()

# Database setup
conn = sqlite3.connect('bot.db')
cursor = conn.cursor()

cursor.execute('''
    CREATE TABLE IF NOT EXISTS users (
        user_id INTEGER PRIMARY KEY,
        username TEXT,
        language TEXT,
        utm_source TEXT,
        step INTEGER DEFAULT 0,
        joined_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
    )
''')
cursor.execute('''
    CREATE TABLE IF NOT EXISTS applications (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        user_id INTEGER,
        full_name TEXT,
        age_city TEXT,
        social_link TEXT,
        photo_file_id TEXT,
        video_file_id TEXT,
        status TEXT DEFAULT 'pending',
        created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
        FOREIGN KEY(user_id) REFERENCES users(user_id)
    )
''')
conn.commit()

# States
class Form(StatesGroup):
    language = State()
    agreement = State()
    full_name = State()
    age_city = State()
    social_link = State()
    photo = State()
    video = State()

# Texts
TEXTS = {
    'ru': {
        'welcome': 'Добро пожаловать в кастинг Dinay Brandface! Пожалуйста, ознакомьтесь с условиями и договором: [Ссылка на договор]. Вы согласны?',
        'agree': '✅ Согласен(на)',
        'ask_name': 'Отлично! Пожалуйста, отправьте ваши Имя и Фамилию текстовым сообщением.',
        'ask_age_city': 'Укажите ваш возраст и город проживания.',
        'ask_social': 'Отправьте ссылку на ваш профиль Instagram или TikTok.',
        'ask_photo': 'Пожалуйста, отправьте скриншот статистики вашего профиля (охваты за 30 дней). Принимаются только фото.',
        'ask_video': 'Отправьте короткое видео (до 30 сек) с ответом на вопрос: «Почему именно вы должны стать лицом Динай?». Принимаются только видео или кружочки.',
        'done': 'Ваша анкета успешно отправлена! Ожидайте результатов.',
        'not_photo': 'Пожалуйста, отправьте именно фото (скриншот статистики).',
        'not_video': 'Пожалуйста, отправьте именно видео или видеосообщение (кружочек).',
        'approved': 'Поздравляем! Ваша заявка одобрена.',
        'rejected': 'К сожалению, ваша заявка отклонена. Спасибо за участие!'
    },
    'uz': {
        'welcome': 'Dinay Brandface kastingiga xush kelibsiz! Iltimos, shartlar va shartnoma bilan tanishib chiqing: [Shartnoma havolasi]. Rozimisiz?',
        'agree': '✅ Roziman',
        'ask_name': 'Ajoyib! Iltimos, Ism va Familiyangizni matnli xabar orqali yuboring.',
        'ask_age_city': 'Yoshingiz va yashash shahringizni kiriting.',
        'ask_social': 'Instagram yoki TikTok profilingiz havolasini yuboring.',
        'ask_photo': 'Iltimos, profilingiz statistikasining skrinshotini yuboring (30 kunlik qamrov). Faqat rasm qabul qilinadi.',
        'ask_video': '«Nima uchun aynan siz Dinay yuzi bo\\'lishingiz kerak?» degan savolga javob berilgan qisqa video (30 soniyagacha) yuboring. Faqat video yoki aylana video qabul qilinadi.',
        'done': 'Sizning anketangiz muvaffaqiyatli yuborildi! Natijalarni kuting.',
        'not_photo': 'Iltimos, faqat rasm (statistika skrinshoti) yuboring.',
        'not_video': 'Iltimos, faqat video yoki video xabar (aylana) yuboring.',
        'approved': 'Tabriklaymiz! Arizangiz qabul qilindi.',
        'rejected': 'Afsuski, arizangiz rad etildi. Ishtirokingiz uchun rahmat!'
    }
}

@dp.message(CommandStart())
async def command_start(message: types.Message, command: CommandObject, state: FSMContext):
    utm = command.args or ""
    user_id = message.from_user.id
    username = message.from_user.username or ""
    
    cursor.execute('INSERT OR IGNORE INTO users (user_id, username, utm_source, step) VALUES (?, ?, ?, 1)', 
                   (user_id, username, utm))
    cursor.execute('UPDATE users SET step = 1 WHERE user_id = ?', (user_id,))
    conn.commit()
    
    keyboard = InlineKeyboardMarkup(inline_keyboard=[
        [InlineKeyboardButton(text="🇷🇺 Русский", callback_data="lang_ru")],
        [InlineKeyboardButton(text="🇺🇿 O'zbekcha", callback_data="lang_uz")]
    ])
    
    await state.set_state(Form.language)
    await message.answer("Выберите язык / Tilni tanlang:", reply_markup=keyboard)

@dp.callback_query(Form.language, F.data.in_(["lang_ru", "lang_uz"]))
async def process_language(callback: types.CallbackQuery, state: FSMContext):
    lang = "ru" if callback.data == "lang_ru" else "uz"
    await state.update_data(language=lang)
    
    cursor.execute('UPDATE users SET language = ?, step = 2 WHERE user_id = ?', (lang, callback.from_user.id))
    conn.commit()
    
    keyboard = InlineKeyboardMarkup(inline_keyboard=[
        [InlineKeyboardButton(text=TEXTS[lang]['agree'], callback_data="agree")]
    ])
    
    await state.set_state(Form.agreement)
    await callback.message.edit_text(TEXTS[lang]['welcome'], reply_markup=keyboard)

@dp.callback_query(Form.agreement, F.data == "agree")
async def process_agreement(callback: types.CallbackQuery, state: FSMContext):
    data = await state.get_data()
    lang = data.get('language', 'ru')
    
    cursor.execute('UPDATE users SET step = 3 WHERE user_id = ?', (callback.from_user.id,))
    conn.commit()
    
    await state.set_state(Form.full_name)
    await callback.message.edit_text(TEXTS[lang]['ask_name'])

@dp.message(Form.full_name, F.text)
async def process_name(message: types.Message, state: FSMContext):
    await state.update_data(full_name=message.text)
    data = await state.get_data()
    lang = data.get('language', 'ru')
    
    cursor.execute('UPDATE users SET step = 4 WHERE user_id = ?', (message.from_user.id,))
    conn.commit()
    
    await state.set_state(Form.age_city)
    await message.answer(TEXTS[lang]['ask_age_city'])

@dp.message(Form.age_city, F.text)
async def process_age_city(message: types.Message, state: FSMContext):
    await state.update_data(age_city=message.text)
    data = await state.get_data()
    lang = data.get('language', 'ru')
    
    cursor.execute('UPDATE users SET step = 5 WHERE user_id = ?', (message.from_user.id,))
    conn.commit()
    
    await state.set_state(Form.social_link)
    await message.answer(TEXTS[lang]['ask_social'])

@dp.message(Form.social_link, F.text)
async def process_social(message: types.Message, state: FSMContext):
    await state.update_data(social_link=message.text)
    data = await state.get_data()
    lang = data.get('language', 'ru')
    
    cursor.execute('UPDATE users SET step = 6 WHERE user_id = ?', (message.from_user.id,))
    conn.commit()
    
    await state.set_state(Form.photo)
    await message.answer(TEXTS[lang]['ask_photo'])

@dp.message(Form.photo, F.photo)
async def process_photo(message: types.Message, state: FSMContext):
    await state.update_data(photo_file_id=message.photo[-1].file_id)
    data = await state.get_data()
    lang = data.get('language', 'ru')
    
    cursor.execute('UPDATE users SET step = 7 WHERE user_id = ?', (message.from_user.id,))
    conn.commit()
    
    await state.set_state(Form.video)
    await message.answer(TEXTS[lang]['ask_video'])

@dp.message(Form.photo)
async def process_photo_invalid(message: types.Message, state: FSMContext):
    data = await state.get_data()
    lang = data.get('language', 'ru')
    await message.answer(TEXTS[lang]['not_photo'])

@dp.message(Form.video, F.video | F.video_note)
async def process_video(message: types.Message, state: FSMContext):
    video_id = message.video.file_id if message.video else message.video_note.file_id
    await state.update_data(video_file_id=video_id)
    data = await state.get_data()
    lang = data.get('language', 'ru')
    
    cursor.execute('UPDATE users SET step = 8 WHERE user_id = ?', (message.from_user.id,))
    
    cursor.execute('''
        INSERT INTO applications (user_id, full_name, age_city, social_link, photo_file_id, video_file_id)
        VALUES (?, ?, ?, ?, ?, ?)
    ''', (
        message.from_user.id,
        data['full_name'],
        data['age_city'],
        data['social_link'],
        data['photo_file_id'],
        video_id
    ))
    conn.commit()
    
    app_id = cursor.lastrowid
    
    await state.clear()
    await message.answer(TEXTS[lang]['done'])
    
    # Notify admins
    for admin_id in ADMIN_IDS:
        try:
            caption = f"Новая анкета #{app_id}\\nФИО: {data['full_name']}\\nВозраст/Город: {data['age_city']}\\nСсылка: {data['social_link']}"
            keyboard = InlineKeyboardMarkup(inline_keyboard=[
                [
                    InlineKeyboardButton(text="✅ Одобрить", callback_data=f"approve_{app_id}"),
                    InlineKeyboardButton(text="❌ Отклонить", callback_data=f"reject_{app_id}")
                ]
            ])
            
            # Send media group (not fully supported with inline keyboards in one message, so we send video with caption)
            await bot.send_photo(admin_id, data['photo_file_id'], caption="Статистика профиля")
            await bot.send_video(admin_id, video_id, caption=caption, reply_markup=keyboard)
        except Exception as e:
            logging.error(f"Failed to send to admin {admin_id}: {e}")

@dp.message(Form.video)
async def process_video_invalid(message: types.Message, state: FSMContext):
    data = await state.get_data()
    lang = data.get('language', 'ru')
    await message.answer(TEXTS[lang]['not_video'])

@dp.callback_query(F.data.startswith("approve_") | F.data.startswith("reject_"))
async def process_admin_decision(callback: types.CallbackQuery):
    if callback.from_user.id not in ADMIN_IDS:
        return
        
    action, app_id = callback.data.split("_")
    status = "approved" if action == "approve" else "rejected"
    
    cursor.execute('UPDATE applications SET status = ? WHERE id = ?', (status, app_id))
    cursor.execute('SELECT user_id FROM applications WHERE id = ?', (app_id,))
    user_id = cursor.fetchone()[0]
    conn.commit()
    
    cursor.execute('SELECT language FROM users WHERE user_id = ?', (user_id,))
    lang = cursor.fetchone()[0] or 'ru'
    
    # Update admin message
    await callback.message.edit_caption(
        caption=f"{callback.message.caption}\\n\\n<b>Статус: {'ОДОБРЕНО' if status == 'approved' else 'ОТКЛОНЕНО'}</b>",
        parse_mode="HTML"
    )
    
    # Notify user
    try:
        await bot.send_message(user_id, TEXTS[lang][status])
    except Exception as e:
        logging.error(f"Failed to notify user {user_id}: {e}")

async def main():
    await dp.start_polling(bot)

if __name__ == "__main__":
    asyncio.run(main())

File: server/bot.ts
import { Bot, Context, session, SessionFlavor, InputFile } from 'grammy';
import { getUser, updateUser, createApplication, getApplications, getBotTexts, updateBotTexts } from './db.js';

interface SessionData {
  language?: 'ru' | 'uz';
  step: number;
  fullName?: string;
  ageCity?: string;
  socialLink?: string;
  langProficiency?: string;
  videoRuId?: string;
  videoUzId?: string;
}

type MyContext = Context & SessionFlavor<SessionData>;

export let bot: Bot<MyContext> | null = null;
let isRunning = false;

let isStarting = false;

export function isBotRunning() {
  return isRunning || isStarting;
}

export async function stopBot() {
  if (bot) {
    try {
      await bot.stop();
    } catch (e) {
      console.error('Error stopping bot:', e);
    }
    isRunning = false;
    bot = null;
  }
}

const defaultTexts = {
  ru: {
    welcome: '👋 Здравствуйте, <b>{name}</b>!\n\nДобро пожаловать на кастинг <b>Dinay Brandface</b>! 🌟\n\nПеред началом, пожалуйста, ознакомьтесь с публичной офертой.\n\nГотовы заполнить анкету?',
    agree: '✅ Начать',
    ask_name: 'Отлично! 🎉\n\nДля начала, пожалуйста, напишите ваши <b>Имя и Фамилию</b>.',
    ask_age_city: 'Супер! Укажите ваш <b>возраст</b> и <b>город проживания</b>.\n\n<i>Пример: 22, Ташкент</i>',
    ask_social: 'Пожалуйста, отправьте ссылку на ваш профиль в <b>Instagram</b> или <b>TikTok</b>. 📱',
    ask_lang_proficiency: '💬 <b>Владеете ли вы русским и узбекским языками?</b>\n\nЛицо бренда Dinay должно свободно общаться на обоих языках. Пожалуйста, оцените ваш уровень владения русским и узбекским языками. 🗣️',
    ask_video_uz: 'Спасибо! Чтобы мы могли оценить вашу харизму и подачу, отправьте короткое <b>видеосообщение (кружочек) на УЗБЕКСКОМ языке</b> (до 30 секунд).\n\nОтветьте на вопрос:\n💬 <i>«Почему именно вы должны стать лицом бренда Dinay?»</i>\n\n🎥 <i>Принимаются только видео или кружочки.</i>',
    ask_video_ru: 'Отлично! А теперь запишите такое же <b>видеосообщение (кружочек) на РУССКОМ языке</b> (до 30 секунд).\n\nОтветьте на тот же вопрос:\n💬 <i>«Почему именно вы должны стать лицом бренда Dinay?»</i>\n\n🎥 <i>Принимаются только видео или кружочки.</i>',
    done: '🎉 Ваша анкета успешно принята!\n\nБлагодарим за интерес к бренду Dinay 💚\nЕсли ваша кандидатура пройдет отбор, наша команда обязательно свяжется с вами.',
    not_video: '❌ Пожалуйста, отправьте <b>видео</b> или <b>видеосообщение (кружочек)</b>.',
    approved: '🎉 <b>Поздравляем, {name}!</b>\n\nВаша заявка на участие в кастинге <b>Dinay Brandface</b> была <b>одобрена</b>! 🥳 Мы скоро свяжемся с вами для дальнейших шагов.',
    rejected: '😔 <b>Здравствуйте, {name}.</b>\n\nК сожалению, ваша заявка была отклонена. Спасибо за проявленный интерес и участие! Желаем успехов в будущем. 🌟'
  },
  uz: {
    welcome: '👋 Assalomu alaykum, <b>{name}</b>!\n\n<b>Dinay Brandface</b> kastingiga xush kelibsiz! 🌟\n\nBoshlashdan oldin, iltimos, ommaviy oferta bilan tanishib chiqing.\n\nAnketani to\'ldirishga tayyormisiz?',
    agree: '✅ Boshlash',
    ask_name: 'Ajoyib! 🎉\n\nBoshlash uchun, iltimos, <b>Ism va Familiyangizni</b> yozib yuboring.',
    ask_age_city: 'Super! Endi <b>yoshingizni</b> va <b>yashash shahringizni</b> kiriting.\n\n<i>Misol: 22, Toshkent</i>',
    ask_social: 'Iltimos, <b>Instagram</b> yoki <b>TikTok</b> profilingiz havolasini yuboring. 📱',
    ask_lang_proficiency: '💬 <b>Rus va o\'zbek tillarini bilasizmi?</b>\n\nDinay brendining yuzi ikkala tilda ham erkin muloqot qila olishi kerak. Iltimos, rus va o\'zbek tillarini qay darajada bilishingizni yozib yuboring. 🗣️',
    ask_video_uz: 'Rahmat! Xarizmangiz va nutqingizni baholashimiz uchun, qisqa <b>video xabar (aylana) O\'ZBEK tilida</b> (30 soniyagacha) yuboring.\n\nQuyidagi savolga javob bering:\n💬 <i>«Nima uchun aynan siz Dinay brendining yuzi bo\'lishingiz kerak?»</i>\n\n🎥 <i>Faqat video yoki aylana video qabul qilinadi.</i>',
    ask_video_ru: 'Ajoyib! Endi xuddi shunday <b>video xabarni (aylana) RUS tilida</b> (30 soniyagacha) yozib yuboring.\n\nXuddi shu savolga javob bering:\n💬 <i>«Nima uchun aynan siz Dinay brendining yuzi bo\'lishingiz kerak?»</i>\n\n🎥 <i>Faqat video yoki aylana video qabul qilinadi.</i>',
    done: '🎉 Sizning anketangiz muvaffaqiyatli qabul qilindi!\n\nDinay brendiga bo\'lgan qiziqishingiz uchun rahmat 💚\nAgar nomzodingiz saralashdan o\'tsa, jamoamiz siz bilan albatta bog\'lanadi.',
    not_video: '❌ Iltimos, <b>video</b> yoki <b>video xabar (aylana)</b> yuboring.',
    approved: '🎉 <b>Tabriklaymiz, {name}!</b>\n\nSizning <b>Dinay Brandface</b> kastingidagi arizangiz <b>qabul qilindi</b>! 🥳 Keyingi qadamlar uchun tez orada siz bilan bog\'lanamiz.',
    rejected: '😔 <b>Assalomu alaykum, {name}.</b>\n\nAfsuski, arizangiz rad etildi. Qiziqish bildirganingiz va ishtirokingiz uchun rahmat! Kelgusidagi ishlaringizda muvaffaqiyatlar tilaymiz. 🌟'
  }
};

let currentTexts: any = null;

export async function reloadTexts() {
  const dbTexts = await getBotTexts();
  currentTexts = {
    ...defaultTexts,
    offer_file_id: dbTexts?.offer_file_id
  };
}

function getText(lang: 'ru' | 'uz', key: string, params: Record<string, string> = {}): string {
  if (!currentTexts) currentTexts = defaultTexts;
  let text = currentTexts[lang]?.[key];
  if (!text || text.trim() === '') {
    text = defaultTexts[lang]?.[key as keyof typeof defaultTexts['ru']] || '';
  }
  for (const [k, v] of Object.entries(params)) {
    text = text.replace(new RegExp(`{${k}}`, 'g'), v);
  }
  return text;
}

async function safeReply(ctx: any, text: string, options: any = {}) {
  try {
    return await ctx.reply(text, options);
  } catch (e) {
    console.error('Failed to send message with HTML, falling back to plain text:', e);
    const plainText = text.replace(/<[^>]*>?/gm, '');
    const safeOptions = { ...options };
    delete safeOptions.parse_mode;
    return await ctx.reply(plainText, safeOptions);
  }
}

export async function initBot(token: string) {
  if (isStarting) {
    console.log('Bot is already starting...');
    return;
  }
  isStarting = true;

  try {
    await stopBot();
    await reloadTexts();
    
    bot = new Bot<MyContext>(token);
    
    bot.use(session({ initial: () => ({ step: 0 }) }));
    
    // Восстанавливаем сессию из базы данных, если она пустая (например, после перезапуска сервера)
    bot.use(async (ctx, next) => {
      if (ctx.from?.id) {
        const user = await getUser(ctx.from.id) as any;
        if (user && ctx.session.step === 0 && user.step > 0) {
          ctx.session.step = user.step;
          ctx.session.language = user.language || user.lang || 'ru';
          ctx.session.fullName = user.full_name;
          ctx.session.ageCity = user.age_city;
          ctx.session.socialLink = user.social_link;
          ctx.session.langProficiency = user.lang_proficiency;
          ctx.session.videoRuId = user.video_ru_id;
          ctx.session.videoUzId = user.video_uz_id;
        }
      }
      await next();
    });
  
  bot.command('setoffer', async (ctx) => {
    // Check if admin
    const groupId = '-5208437302';
    const allowedAdmins = [768567874]; // Ваш Telegram ID
    
    if (!allowedAdmins.includes(ctx.from.id)) {
      try {
        const chatMember = await ctx.api.getChatMember(groupId, ctx.from.id);
        if (chatMember.status !== 'creator' && chatMember.status !== 'administrator') {
          await safeReply(ctx, '⛔️ У вас нет прав администратора для выполнения этой команды.');
          return;
        }
      } catch (e) {
        console.error('Admin check error:', e);
        await safeReply(ctx, '⛔️ У вас нет прав для выполнения этой команды.');
        return;
      }
    }

    let document = ctx.message?.document;
    
    // Check if it's a reply to a document
    if (!document && ctx.message?.reply_to_message?.document) {
      document = ctx.message.reply_to_message.document;
    }

    if (!document) {
      await safeReply(ctx, 
        '⚠️ <b>Как обновить оферту:</b>\n\n' +
        '1. Отправьте файл (PDF или другой документ) в этот чат.\n' +
        '2. В поле "Подпись" (Caption) напишите команду <code>/setoffer</code>\n' +
        '<b>ИЛИ</b>\n' +
        'Ответьте (Reply) на уже отправленный файл командой <code>/setoffer</code>',
        { parse_mode: 'HTML' }
      );
      return;
    }

    const fileId = document.file_id;
    let texts = await getBotTexts() || defaultTexts;
    texts.offer_file_id = fileId;
    
    await updateBotTexts(texts);
    await reloadTexts();
    
    await safeReply(ctx, '✅ Публичная оферта успешно обновлена!');
  });

  bot.command('start', async (ctx) => {
    const utm = ctx.match || '';
    const userId = ctx.from?.id;
    const username = ctx.from?.username || '';
    
    if (userId) {
      const existingUser = await getUser(userId);
      if (!existingUser) {
        await updateUser(userId, {
          username,
          utm_source: utm,
          step: 1,
          joined_at: new Date().toISOString()
        });
      } else {
        await updateUser(userId, { step: 1 });
      }
    }
    
    ctx.session.step = 1;
    
    await safeReply(ctx, 'Выберите язык / Tilni tanlang:', {
      reply_markup: {
        inline_keyboard: [
          [{ text: '🇷🇺 Русский', callback_data: 'lang_ru' }],
          [{ text: "🇺🇿 O'zbekcha", callback_data: 'lang_uz' }]
        ]
      }
    });
  });
  
  bot.on('callback_query:data', async (ctx) => {
    const data = ctx.callbackQuery.data;
    const userId = ctx.from.id;
    
    if (data === 'lang_ru' || data === 'lang_uz') {
      const lang = data === 'lang_ru' ? 'ru' : 'uz';
      ctx.session.language = lang;
      ctx.session.step = 2;
      
      const userName = [ctx.from.first_name, ctx.from.last_name].filter(Boolean).join(' ') || 'Участник';
      const safeName = userName.replace(/</g, '&lt;').replace(/>/g, '&gt;');
      
      await updateUser(userId, { language: lang, step: 2 });
      
      await ctx.answerCallbackQuery().catch(() => {});
      await ctx.deleteMessage().catch(() => {});
      
      // Send offer file if exists
      if (currentTexts?.offer_file_id) {
        await ctx.api.sendDocument(userId, currentTexts.offer_file_id, {
          caption: '📄 Публичная оферта / Ommaviy oferta'
        }).catch(console.error);
      }
      
      try {
        await safeReply(ctx, 
          getText(lang, 'welcome', { name: safeName }),
          {
            parse_mode: 'HTML',
            reply_markup: {
              inline_keyboard: [
                [{ text: getText(lang, 'agree'), callback_data: 'agree' }]
              ]
            }
          }
        );
      } catch (e) {
        console.error('Failed to send welcome message with HTML, falling back to plain text:', e);
        await safeReply(ctx, 
          getText(lang, 'welcome', { name: safeName }).replace(/<[^>]*>?/gm, ''),
          {
            reply_markup: {
              inline_keyboard: [
                [{ text: getText(lang, 'agree'), callback_data: 'agree' }]
              ]
            }
          }
        );
      }
    } else if (data === 'agree') {
      const lang = ctx.session.language || 'ru';
      ctx.session.step = 3;
      await updateUser(userId, { step: 3 });
      
      await ctx.answerCallbackQuery().catch(() => {});
      // Убираем кнопку "Согласен", чтобы нельзя было нажать дважды
      await ctx.editMessageReplyMarkup({ reply_markup: { inline_keyboard: [] } }).catch(() => {});
      // Отправляем следующий вопрос новым сообщением
      await safeReply(ctx, getText(lang, 'ask_name'), { parse_mode: 'HTML' });
    }
  });

  bot.on('message:text', async (ctx) => {
    const step = ctx.session.step;
    const lang = ctx.session.language || 'ru';
    const userId = ctx.from.id;
    
    if (step === 3) {
      ctx.session.fullName = ctx.message.text;
      ctx.session.step = 4;
      await updateUser(userId, { step: 4, full_name: ctx.message.text });
      await safeReply(ctx, getText(lang, 'ask_age_city'), { parse_mode: 'HTML' });
    } else if (step === 4) {
      ctx.session.ageCity = ctx.message.text;
      ctx.session.step = 5;
      await updateUser(userId, { step: 5, age_city: ctx.message.text });
      await safeReply(ctx, getText(lang, 'ask_social'), { parse_mode: 'HTML' });
    } else if (step === 5) {
      ctx.session.socialLink = ctx.message.text;
      ctx.session.step = 6;
      await updateUser(userId, { step: 6, social_link: ctx.message.text });
      await safeReply(ctx, getText(lang, 'ask_lang_proficiency'), { parse_mode: 'HTML' });
    } else if (step === 6) {
      ctx.session.langProficiency = ctx.message.text;
      ctx.session.step = 7;
      await updateUser(userId, { step: 7, lang_proficiency: ctx.message.text });
      await safeReply(ctx, getText(lang, 'ask_video_uz'), { parse_mode: 'HTML' });
    }
  });

  bot.on(['message:video', 'message:video_note'], async (ctx) => {
    const step = ctx.session.step;
    const lang = ctx.session.language || 'ru';
    const userId = ctx.from.id;
    
    if (step === 7) {
      ctx.session.videoUzId = ctx.message.video?.file_id || ctx.message.video_note?.file_id;
      ctx.session.step = 8;
      await updateUser(userId, { step: 8, video_uz_id: ctx.session.videoUzId });
      await safeReply(ctx, getText(lang, 'ask_video_ru'), { parse_mode: 'HTML' });
    } else if (step === 8) {
      ctx.session.videoRuId = ctx.message.video?.file_id || ctx.message.video_note?.file_id;
      ctx.session.step = 9;
      await updateUser(userId, { step: 9, video_ru_id: ctx.session.videoRuId });
      
      await createApplication({
        user_id: userId,
        full_name: ctx.session.fullName,
        age_city: ctx.session.ageCity,
        social_link: ctx.session.socialLink,
        lang_proficiency: ctx.session.langProficiency,
        video_ru_id: ctx.session.videoRuId,
        video_uz_id: ctx.session.videoUzId
      });
      
      // Отправка уведомления в группу
      try {
        const groupId = '-5208437302';
        const username = ctx.from?.username;
        const userLink = username ? `@${username}` : `<a href="tg://user?id=${userId}">Пользователь (ID: ${userId})</a>`;
        
        const notifyText = `🆕 <b>Новая заявка!</b>\n\n` +
          `👤 <b>ФИО:</b> ${ctx.session.fullName}\n` +
          `🏙 <b>Возраст/Город:</b> ${ctx.session.ageCity}\n` +
          `📱 <b>Соцсеть:</b> ${ctx.session.socialLink}\n` +
          `🗣 <b>Знание языков:</b> ${ctx.session.langProficiency}\n` +
          `💬 <b>Telegram:</b> ${userLink}`;

        await ctx.api.sendMessage(groupId, notifyText, { parse_mode: 'HTML' });
        
        // Отправляем видео или кружочек
        if (ctx.session.videoUzId) {
          await ctx.api.sendMessage(groupId, '🎥 Видео на узбекском:');
          try {
            await ctx.api.sendVideoNote(groupId, ctx.session.videoUzId);
          } catch (e) {
            await ctx.api.sendVideo(groupId, ctx.session.videoUzId).catch(console.error);
          }
        }
        if (ctx.session.videoRuId) {
          await ctx.api.sendMessage(groupId, '🎥 Видео на русском:');
          try {
            await ctx.api.sendVideoNote(groupId, ctx.session.videoRuId);
          } catch (e) {
            await ctx.api.sendVideo(groupId, ctx.session.videoRuId).catch(console.error);
          }
        }
      } catch (err) {
        console.error('Ошибка при отправке в группу:', err);
      }
      
      await safeReply(ctx, getText(lang, 'done'), { parse_mode: 'HTML' });
    }
  });
  
  bot.on('message', async (ctx) => {
    const step = ctx.session.step;
    const lang = ctx.session.language || 'ru';
    
    if ((step === 7 || step === 8) && !ctx.message.video && !ctx.message.video_note) {
      await safeReply(ctx, getText(lang, 'not_video'), { parse_mode: 'HTML' });
    }
  });
  
  bot.catch((err) => {
    console.error('Bot Error:', err);
  });
  
  bot.start({
    onStart: (botInfo) => {
      console.log(`Bot started as @${botInfo.username}`);
      isRunning = true;
      isStarting = false;
    }
  }).catch((err) => {
    console.error('Failed to start polling:', err);
    isRunning = false;
    isStarting = false;
  });
  } catch (e) {
    console.error('Error during bot initialization:', e);
    isStarting = false;
    throw e;
  }
}

export async function getTelegramFileUrl(fileId: string): Promise<string> {
  if (!bot || !isRunning) throw new Error('Bot is not running');
  const file = await bot.api.getFile(fileId);
  return `https://api.telegram.org/file/bot${bot.token}/${file.file_path}`;
}

export async function notifyUser(userId: number, status: 'approved' | 'rejected') {
  if (!bot || !isRunning) return;
  
  try {
    const user = await getUser(userId) as any;
    const apps = await getApplications();
    const app = apps.find(a => a.user_id === userId);
    
    const lang = (user?.lang || user?.language || 'ru') as 'ru' | 'uz';
    const safeName = (app?.full_name || 'Участник').replace(/</g, '&lt;').replace(/>/g, '&gt;');
    
    await bot.api.sendMessage(userId, getText(lang, status, { name: safeName }), { parse_mode: 'HTML' });
  } catch (e) {
    console.error('Failed to notify user:', e);
  }
}
